### PR TITLE
satellite/payments: invoice creation

### PR DESCRIPTION
**Pull request by rikysya**
What:  add functionality for creating stripe invoices for `satellite` users. Add private `RPC` endpoint that exposes invoice related functionality to `satellite` owner

Why: to add ability for `satellite` owner to create invoice for it's users.

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [x] NEW: Are there any Satellite database migrations? Are they forwards _and_ backwards compatible? 
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?
 - [x] Do the database access patterns make sense?

### CODE REVIEW

File: cmd/inspector/main.go [hunk 1]
```diff
@@ -423,14 +446,104 @@ func sortSegments(segments []*pb.IrreparableSegment) map[string][]*pb.Irreparabl
...
+	}
+
+	defer func() { err = errs.Combine(err, i.Close()) }()
+
+	_, err = i.paymentsClient.CreateInvoices(context.Background(), &pb.CreateInvoicesRequest{})
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("successfully created invoices")
```
Comment (crawter) [inline cmd/inspector/main.go]: invoices are successfully created

File: satellite/api.go [hunk 2]
```diff
@@ -154,6 +155,9 @@ func NewAPI(log *zap.Logger, full *identity.FullIdentity, db DB, pointerDB metai
 		log.Debug("Satellite API Process starting listener and server")
 		sc := config.Server
 
+		log.Debug(fmt.Sprintf("Satellite API Public %s", sc.Address))
```
Comment (NikolaiYurchenko) [inline satellite/api.go]: Remove debug code

File: satellite/payments/stripecoinpayments/service.go [hunk 3]
```diff
@@ -90,7 +92,8 @@ func (service *Service) updateTransactionsLoop(ctx context.Context) (err error)
 			return err
 		}
 
-		txsPage, err = service.transactionsDB.ListPending(ctx, txsPage.NextOffset, limit, before)
+		txsPage, err = service.db.Transactions().ListPending(ctx, txsPage.NextOffset, limit, before)
+
```
Comment (NikolaiYurchenko) [inline satellite/payments/stripecoinpayments/service.go]: Remove empty line

File: satellite/payments/stripecoinpayments/service.go [hunk 4]
```diff
@@ -216,3 +219,243 @@ func (service *Service) applyTransactionBalance(ctx context.Context, tx Transact
...
+	if err = service.createProjectRecords(ctx, projsPage.Projects, start, end); err != nil {
+		return Error.Wrap(err)
+	}
+
+	for projsPage.Next {
+		if err = ctx.Err(); err != nil {
+			return Error.Wrap(err)
+		}
+
+		projsPage, err = service.projectsDB.List(ctx, 0, limit, end)
```
Comment (NikolaiYurchenko) [inline satellite/payments/stripecoinpayments/service.go]: projsPage.NextOffset

File: satellite/satellitedb/customers.go [hunk 5]
```diff
@@ -16,6 +18,8 @@ import (
 var _ stripecoinpayments.CustomersDB = (*customers)(nil)
 
 // customers is an implementation of stripecoinpayments.CustomersDB.
+//
+// architecture: Database
```
Comment (crawter) [inline satellite/satellitedb/customers.go]: not sure that this annotation is needed

File: satellite/payments/stripecoinpayments/service.go [hunk 6]
```diff
@@ -216,3 +219,243 @@ func (service *Service) applyTransactionBalance(ctx context.Context, tx Transact
...
+	if err = service.applyProjectRecords(ctx, recordsPage.Records); err != nil {
+		return Error.Wrap(err)
+	}
+
+	for recordsPage.Next {
+		if err = ctx.Err(); err != nil {
+			return Error.Wrap(err)
+		}
+
+		recordsPage, err = service.db.ProjectRecords().ListUnapplied(ctx, 0, limit, before)
```
Comment (NikolaiYurchenko) [inline satellite/payments/stripecoinpayments/service.go]: recordsPage.NextOffset

File: satellite/satellitedb/invoiceprojectrecords.go [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+// Int returns intent state as int.
+func (intent invoiceProjectRecordState) Int() int {
+	return int(intent)
+}
+
+// invoiceProjectRecords is stripecoinpayments project records DB.
+//
+// architecture: Database
```
Comment (crawter) [inline satellite/satellitedb/invoiceprojectrecords.go]: not sure that this annotations in needed

File: satellite/payments/stripecoinpayments/service.go [hunk 8]
```diff
@@ -216,3 +219,243 @@ func (service *Service) applyTransactionBalance(ctx context.Context, tx Transact
...
+		Period: &stripe.InvoiceItemPeriodParams{
+			End:   stripe.Int64(record.PeriodEnd.Unix()),
+			Start: stripe.Int64(record.PeriodStart.Unix()),
+		},
+	}
+
+	projectItem.AddMetadata("projectID", record.ProjectID.String())
+
+	_, err = service.stripeClient.InvoiceItems.New(projectItem)
+	return
```
Comment (NikolaiYurchenko) [inline satellite/payments/stripecoinpayments/service.go]: return err

File: satellite/payments/stripecoinpayments/service.go [hunk 9]
```diff
@@ -216,3 +219,243 @@ func (service *Service) applyTransactionBalance(ctx context.Context, tx Transact
...
+			return Error.Wrap(err)
+		}
+	}
+
+	for cusPage.Next {
+		if err = ctx.Err(); err != nil {
+			return Error.Wrap(err)
+		}
+
+		cusPage, err = service.db.Customers().List(ctx, 0, limit, before)
```
Comment (NikolaiYurchenko) [inline satellite/payments/stripecoinpayments/service.go]: cusPage.NextOffset

File: satellite/satellitedb/projectaccounting.go [hunk 10]
```diff
@@ -16,6 +16,9 @@ import (
 	dbx "storj.io/storj/satellite/satellitedb/dbx"
 )
 
+// ensure that ProjectAccounting implements accounting.ProjectAccounting.
```
Comment (NikolaiYurchenko) [inline satellite/satellitedb/projectaccounting.go]: Can be removed

File: cmd/inspector/main.go [hunk 11]
```diff
@@ -79,15 +80,36 @@ var (
...
-	identity      *identity.FullIdentity
-	overlayclient rpc.OverlayInspectorClient
-	irrdbclient   rpc.IrreparableInspectorClient
-	healthclient  rpc.HealthInspectorClient
+	conn           *rpc.Conn
+	identity       *identity.FullIdentity
+	overlayclient  rpc.OverlayInspectorClient
+	irrdbclient    rpc.IrreparableInspectorClient
+	healthclient   rpc.HealthInspectorClient
+	paymentsClient rpc.PaymentsClient
```
Comment (egonelbre) [inline cmd/inspector/main.go]: Why would this be part of an inspector? Inspector is all about inspecting not changing the state.

File: satellite/api.go [hunk 12]
```diff
@@ -108,8 +109,8 @@ type API struct {
 	}
 
 	Payments struct {
-		Accounts payments.Accounts
-		Clearing payments.Clearing
+		Accounts  payments.Accounts
+		Inspector *stripecoinpayments.Endpoint
```
Comment (egonelbre) [inline satellite/api.go]: It's really not about inspecting, it's about other things as well.

File: cmd/inspector/main.go (hunk 11 above)
Comment (rikysya) [reply cmd/inspector/main.go]: Once @jtolds  told me that we should place it there

File: satellite/api.go (hunk 12 above)
Comment (rikysya) [reply satellite/api.go]: I called it like this so it is easier to understand where we should invoke this

File: cmd/inspector/main.go (hunk 11 above)
Comment (egonelbre) [reply cmd/inspector/main.go]: When I use inspector I would expect that when I call anything on it, then it's safe to do concurrently.

File: cmd/inspector/main.go [hunk 13]
```diff
@@ -423,14 +446,104 @@ func sortSegments(segments []*pb.IrreparableSegment) map[string][]*pb.Irreparabl
...
+		return err
+	}
+
+	fmt.Println("successfully created invoices")
+	return nil
+}
+
+// parseDateString parses provided date string and returns corresponding time.Time.
+func parseDateString(s string) (time.Time, error) {
+	values := strings.Split(s, "/")
```
Comment (egonelbre) [inline cmd/inspector/main.go]: `return time.ParseInLocation("2006/01", s, time.UTC)

File: satellite/payments/stripecoinpayments/endpoint.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+	return &Endpoint{service: service}
+}
+
+// PrepareInvoiceRecords creates project invoice records for all satellite projects.
+func (endpoint *Endpoint) PrepareInvoiceRecords(ctx context.Context, req *pb.PrepareInvoiceRecordsRequest) (_ *pb.PrepareInvoiceRecordsResponse, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	err = endpoint.service.PrepareInvoiceProjectRecords(ctx, req.Period)
+	if err != nil {
+		return nil, err
```
Comment (egonelbre) [inline satellite/payments/stripecoinpayments/endpoint.go]: return rpcstatus codes

File: satellite/payments/stripecoinpayments/service.go [hunk 15]
```diff
@@ -216,3 +218,243 @@ func (service *Service) applyTransactionBalance(ctx context.Context, tx Transact
 	_, err = service.stripeClient.CustomerBalanceTransactions.New(params)
 	return err
 }
+
+// PrepareInvoiceProjectRecords iterates through all projects and creates invoice records if
+// none exists.
+func (service *Service) PrepareInvoiceProjectRecords(ctx context.Context, period time.Time) (err error) {
```
Comment (egonelbre) [inline satellite/payments/stripecoinpayments/service.go]: Maybe future change to "year, month int"

**Pull request merged**